### PR TITLE
feat: support RPM archives

### PR DESCRIPTION
**Pull request by knqyf263**
## Description
Add support for RPM files. It's disabled unless `TRIVY_EXPERIMENTAL_RPM_ARCHIVE` is set as it's experimental.

```
$ TRIVY_EXPERIMENTAL_RPM_ARCHIVE=true ./trivy fs ./redhat-7.9-rpms -f cyclonedx
```

<details><summary>Details</summary>
<p>

```
2024-10-02T13:46:05+04:00       INFO    "--format cyclonedx" disables security scanning. Specify "--scanners vuln" explicitly if you want to include vulnerabilities in the "cyclonedx" report.
2024-10-02T13:46:05+04:00       INFO    Detected OS     family="none" version=""
2024-10-02T13:46:05+04:00       INFO    Number of language-specific files       num=0
{
  "$schema": "http://cyclonedx.org/schema/bom-1.6.schema.json",
  "bomFormat": "CycloneDX",
  "specVersion": "1.6",
  "serialNumber": "urn:uuid:ca46e752-2b04-49fc-a237-067b200be2e4",
  "version": 1,
  "metadata": {
    "timestamp": "2024-10-02T09:46:05+00:00",
    "tools": {
      "components": [
        {
          "type": "application",
          "group": "aquasecurity",
          "name": "trivy",
          "version": "0.55.0-49-g981a8b3e9"
        }
      ]
    },
    "component": {
      "bom-ref": "48aad1f0-9af0-445e-afd4-26a44dc2acda",
      "type": "application",
      "name": "/tmp/rpms/redhat-7.9-rpm-for-trivy",
      "properties": [
        {
          "name": "aquasecurity:trivy:SchemaVersion",
          "value": "2"
        }
      ]
    }
  },
  "components": [
    {
      "bom-ref": "4194564a-f44c-4a33-b848-6e570af8c7b5",
      "type": "operating-system",
      "name": "none",
      "properties": [
        {
          "name": "aquasecurity:trivy:Class",
          "value": "os-pkgs"
        },
        {
          "name": "aquasecurity:trivy:Type",
          "value": "none"
        }
      ]
    },
    {
      "bom-ref": "pkg:rpm/redhat/adcli-doc@0.8.2-12.el8",
      "type": "library",
      "supplier": {
        "name": "Red Hat, Inc."
      },
      "name": "adcli-doc",
      "version": "0.8.2-12.el8",
      "licenses": [
        {
          "license": {
            "name": "LGPLv2+"
          }
        }
      ],
```

</p>
</details> 

### TODOs
- [x] Parse RPM files
- [x] Enable/Disable the analyzer 
- [x] Add tests

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [ ] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [x] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

**Comment by knqyf263** [comment]
@DmitriyLewen Some tests are failing due to 429 from GHCR, but it's ready for review.

File: pkg/fanal/analyzer/pkg/rpm/archive.go [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+		ns = "redhat"
+	case strings.Contains(vendor, "fedora"):
+		ns = "fedora"
+	case strings.Contains(vendor, "opensuse"):
+		ns = "opensuse"
+	case strings.Contains(vendor, "suse"):
+		ns = "suse"
+	}
+	return packageurl.NewPackageURL(packageurl.TypeRPM, ns, pkg.Name, utils.FormatVersion(*pkg), nil, "")
+}
```
[the remark this answers is not in the document]
Comment (knqyf263) [reply pkg/fanal/analyzer/pkg/rpm/archive.go:128]: We need to generate PURL here as OS is not detected when scanning RPM files.

[the remark this answers is not in the document]
Comment (DmitriyLewen) [reply pkg/fanal/analyzer/pkg/rpm/archive.go:140]: I agree that we need to generate PURL here.
I am worried about case when we scan image that containing `*.rpm` file, and `OS` and vendor of this file are not equal.
In this case we will use namespace from OS.
(example): `fedora` image contains `socat-1.7.3.2-2.el7.x86_64.rpm` file (for RedHat).
in analyzer we use `redhat` namespace.
But in `applier` we will overwrite purl and use `fedora` namespace.

I meant that perhaps we need to add `purl == nil` check in applier.

File: pkg/fanal/analyzer/pkg/rpm/archive.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+		return types.Package{}, xerrors.Errorf("failed to get licenses: %w", err)
+	}
+	sourceRpm, err := h.GetString(rpmutils.SOURCERPM)
+	if a.unexpectedError(err) != nil {
+		return types.Package{}, xerrors.Errorf("failed to get source rpm: %w", err)
+	}
+	srcName, srcVer, srcRel, err := splitFileName(sourceRpm)
+	if err != nil {
+		a.logger.Debug("Invalid Source RPM Found", log.String("sourcerpm", sourceRpm))
+	}
```
[the remark this answers is not in the document]
Comment (knqyf263) [reply pkg/fanal/analyzer/pkg/rpm/archive.go]: Fixed in https://github.com/aquasecurity/trivy/pull/7628/commits/c1fa209042afc92f93bb40b2b2297ad4a0f62eae

File: pkg/fanal/analyzer/pkg/rpm/archive.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (knqyf263) [reply pkg/fanal/analyzer/pkg/rpm/archive.go:140]: It makes sense. Added https://github.com/aquasecurity/trivy/pull/7628/commits/9c26be0341c7107502b2595e047af69cd84d8f4c

**Comment by DmitriyLewen** [comment]
I found another problem:
we can't detect vulnerabilities if OS is not found and we remove rpm packages:
```
➜ TRIVY_EXPERIMENTAL_RPM_ARCHIVE=true ./trivy -q fs -f json --list-all-pkgs ./pkg/fanal/analyzer/pkg/rpm/testdata/socat-1.7.3.2-2.el7.x86_64.rpm
{
  "SchemaVersion": 2,
  "CreatedAt": "2024-10-02T17:16:45.768306+06:00",
  "ArtifactName": "pkg/fanal/analyzer/pkg/rpm/testdata/socat-1.7.3.2-2.el7.x86_64.rpm",
  "ArtifactType": "filesystem",
  "Metadata": {
    "OS": {
      "Family": "none",
      "Name": ""
    },
    "ImageConfig": {
      "architecture": "",
      "created": "0001-01-01T00:00:00Z",
      "os": "",
      "rootfs": {
        "type": "",
        "diff_ids": null
      },
      "config": {}
    }
  }
}

```

It works only when `vuln` scanner is disable.
```
➜  TRIVY_EXPERIMENTAL_RPM_ARCHIVE=true ./trivy -q fs -f cyclonedx  ./pkg/fanal/analyzer/pkg/rpm/testdata/socat-1.7.3.2-2.el7.x86_64.rpm | grep socat@1.7.3.2-2
      "bom-ref": "pkg:rpm/redhat/socat@1.7.3.2-2.el7",
      "purl": "pkg:rpm/redhat/socat@1.7.3.2-2.el7",
        "pkg:rpm/redhat/socat@1.7.3.2-2.el7"
      "ref": "pkg:rpm/redhat/socat@1.7.3.2-2.el7",

```

See https://github.com/aquasecurity/trivy/blob/4926da79de901fba73819d71845ec0355b68ae0f/pkg/scanner/local/scan.go#L165-L166

**Comment by knqyf263** [comment]
Nice catch. I added a note.
https://github.com/aquasecurity/trivy/pull/7628/commits/ab37b6a4263bc6c5f3a47db307c28a982c9f4531

**Review by DmitriyLewen — APPROVED** [review]
LGTM

**Pull request merged**